flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -421,6 +437,15 @@ private void addStopStateToLooping(final State<T> loopingState) {
 				untilCondition,
 				true);
 
+			if (currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY) &&
+				times.getFrom() != times.getTo()) {
+				if (untilCondition != null) {
+					State<T> sinkStateCopy = copy(sinkState);
+					originalStateMap.put(sinkState.getName(), sinkStateCopy);
```

Comment:
When are the old entries cleared in this map ?
Shall we consider using map which expires entries by TTL ?

Reply:
originalStateMap is used when compiling the NFA and it will be collected after NFA is created and so I think it's unnecessary to clear the entries.